File: internal/engine/pricing.go
package engine

import "math"

// pricing.go — EVE-legal price helpers shared across the engine and the api
// layer. EVE enforces a 4-significant-digit rule on market order prices: at
// magnitude M, the smallest legal step is 10^(M-3). So a price near 12.3M
// steps in 10k increments; near 12.3B in 1M increments; near 12.3 ISK in
// 0.01 increments.
//
// These functions are the single source of truth for "what price would EVE
// actually accept". Reused by:
//   - internal/engine/undercut.go       (Order Desk suggested reprice)
//   - internal/engine/order_desk.go     (Command Center suggested reprice)
//   - internal/api/price_audit.go       (multisell "Import Prices" workflow)
//   - internal/api/hub_allocate.go      (hub allocator sell price)
//   - internal/api/pi_factory.go        (PI output undercut display)

// NextSellUndercut returns the largest EVE-legal price strictly less than
// lowestSell. Snaps to the 4-sig-fig grid at the appropriate magnitude.
// Returns 0 when the input isn't a positive finite number.
func NextSellUndercut(lowestSell float64) float64 {
	if !(lowestSell > 0) || math.IsInf(lowestSell, 0) || math.IsNaN(lowestSell) {
		return 0
	}
	place := priceStep(lowestSell)
	if place <= 0 {
		return 0
	}
	// Snap down to the nearest 4-sig-fig grid.
	floored := math.Floor(lowestSell/place) * place
	if floored < lowestSell {
		return SnapToGrid(floored, place)
	}
	// Already on a valid boundary — step down one place.
	stepped := lowestSell - place
	if stepped <= 0 {
		return 0
	}
	return SnapToGrid(stepped, place)
}

// NextBuyOverbid returns the smallest EVE-legal price strictly greater than
// topBuy. Mirror of NextSellUndercut for the bid side: snaps UP to the next
// 4-sig-fig grid boundary above topBuy; when topBuy is already on a
// boundary, adds one place value. Returns 0 for non-positive input.
//
// Note: as topBuy crosses a power-of-ten boundary (say 999,900 → 1,000,000)
// the place value jumps from 100 → 1000. Snapping up from just under the
// boundary lands on the boundary itself (a valid legal price), then the
// next call uses the larger step. This matches EVE's behaviour.
func NextBuyOverbid(topBuy float64) float64 {
	if !(topBuy > 0) || math.IsInf(topBuy, 0) || math.IsNaN(topBuy) {
		return 0
	}
	place := priceStep(topBuy)
	if place <= 0 {
		return 0
	}
	// Snap up to the nearest 4-sig-fig grid.
	ceiled := math.Ceil(topBuy/place) * place
	if ceiled > topBuy {
		return SnapToGrid(ceiled, place)
	}
	// Already on a valid boundary — step up one place.
	return SnapToGrid(topBuy+place, place)
}

// SnapToGrid rounds value to the nearest multiple of place, scrubbing
// IEEE-754 noise from subsequent add/subtract (e.g. `19.17 - 0.01`
// producing `19.169999...998`). After this the value is a clean multiple
// of place and JSON-encodes as the expected decimal string.
func SnapToGrid(value, place float64) float64 {
	if place <= 0 {
		return value
	}
	return math.Round(value/place) * place
}

// priceStep returns the 4-sig-fig place value at the magnitude of price.
// Kept unexported since the two public helpers cover every documented use
// case and a raw place-value is rarely what a caller wants.
func priceStep(price float64) float64 {
	magnitude := math.Floor(math.Log10(price))
	return math.Pow(10, magnitude-3)
}

File: frontend/src/lib/pricing.ts
// pricing.ts — EVE-legal price helpers. A direct port of
// internal/engine/pricing.go, which is the source of truth for "what price
// would EVE actually accept".
//
// EVE enforces a 4-significant-digit rule on market order prices: at
// magnitude M, the smallest legal step is 10^(M-3). So a price near 12.3M
// steps in 10k increments; near 12.3B in 1M increments; near 12.3 ISK in
// 0.01 increments.
//
// Keep this in lockstep with the Go original. If the two ever disagree, the
// Order Desk's suggested reprice and the price we drop on the clipboard stop
// agreeing about the same book, which is worse than either being wrong alone.

/** The smallest legal price increment at `price`'s magnitude. */
export function priceStep(price: number): number {
  const magnitude = Math.floor(Math.log10(price));
  return Math.pow(10, magnitude - 3);
}

/**
 * Rounds `value` onto the `place`-sized grid, scrubbing IEEE-754 noise
 * (19.17 - 0.01 naïvely yields 19.169999999999998).
 */
export function snapToGrid(value: number, place: number): number {
  if (place <= 0) return value;
  return Math.round(value / place) * place;
}

/**
 * The highest EVE-legal price strictly below `lowestSell` — what you list at
 * to take the top of the sell side. Returns 0 for anything unusable.
 */
export function nextSellUndercut(lowestSell: number): number {
  if (!(lowestSell > 0) || !Number.isFinite(lowestSell)) return 0;
  const place = priceStep(lowestSell);
  if (!(place > 0)) return 0;

  // Snap down to the nearest 4-sig-fig grid.
  const floored = Math.floor(lowestSell / place) * place;
  if (floored < lowestSell) return snapToGrid(floored, place);

  // Already on a valid boundary — step down one place.
  const stepped = lowestSell - place;
  if (stepped <= 0) return 0;
  return snapToGrid(stepped, place);
}

/**
 * The lowest EVE-legal price strictly above `topBuy` — what you bid at to take
 * the top of the buy side. Returns 0 for anything unusable.
 */
export function nextBuyOverbid(topBuy: number): number {
  if (!(topBuy > 0) || !Number.isFinite(topBuy)) return 0;
  const place = priceStep(topBuy);
  if (!(place > 0)) return 0;

  const ceiled = Math.ceil(topBuy / place) * place;
  if (ceiled > topBuy) return snapToGrid(ceiled, place);
  return snapToGrid(topBuy + place, place);
}

/**
 * Renders a grid price with exactly the decimals its step needs, so pasting it
 * into EVE reproduces the number we computed. `toFixed(2)` would round 5.499
 * (a legal sub-10-ISK undercut) up to 5.50 — back above the price we were
 * trying to undercut.
 */
export function formatGridPrice(value: number, place: number): string {
  if (!Number.isFinite(value)) return "";
  const decimals = place > 0 ? Math.min(4, Math.max(0, -Math.floor(Math.log10(place)))) : 2;
  return value.toFixed(decimals);
}
